File: Android-Demo/UiMaster/app/src/main/java/com/wx/uimaster/adapter/base/BaseAdapter.java
package com.wx.uimaster.adapter.base;

import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.wx.uimaster.adapter.listener.OnItemClickListener;
import com.wx.uimaster.adapter.viewholder.BaseViewHolder;

import java.util.ArrayList;
import java.util.List;

public abstract class BaseAdapter<T> extends RecyclerView.Adapter<BaseViewHolder>{

    private OnItemClickListener mListener;

    private List<T> mDatas;

    public abstract int getLayouId();


    public abstract void onBindView(BaseViewHolder holder, T data);

    public BaseAdapter() {
        mDatas = new ArrayList<>();
    }

    public void setOnItemClickListener(OnItemClickListener listener) {
        this.mListener = listener;
    }

    @Override
    public BaseViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        ViewDataBinding binding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()),getLayouId(),parent,false);
        BaseViewHolder viewHolder = new BaseViewHolder(binding.getRoot());
        viewHolder.setBinding(binding);
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(BaseViewHolder holder, final int position) {
        if (mListener != null) {
            holder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    mListener.onItemClick(view,position);
                }
            });
        }
        onBindView(holder,mDatas.get(position));
    }

    @Override
    public int getItemCount() {
        return mDatas.size();
    }

    public void addDatas(List<T> data,boolean isWantAnimator) {
        int start = mDatas.size();
        mDatas.addAll(data);
        int end = mDatas.size()-1;
        if (isWantAnimator) {
            notifyItemRangeInserted(start,end);
        } else {
            notifyDataSetChanged();
        }
    }

    public void addData(T data,boolean isWantAnimator) {
        mDatas.add(data);
        if (isWantAnimator) {
            notifyItemInserted(mDatas.size()-1);
        } else {
            notifyDataSetChanged();
        }
    }

}

File: Android-Demo/NetWork/GlideDemo1/app/src/main/java/com/wx/glidedemo1/adapter/base/BaseAdapter.kt
package com.ppdl.remoteclient.adapter.base

import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.support.annotation.LayoutRes
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.ppdl.remoteclient.adapter.listener.OnItemClickListner
import com.ppdl.remoteclient.adapter.viewholder.BaseViewHolder

/**
 * Created by wangxuan on 2018/1/2.
 */

abstract class BaseAdapter<T>() : RecyclerView.Adapter<BaseViewHolder>() {

    @get: LayoutRes
    abstract val layoutId: Int

    abstract fun onBindView(holder:BaseViewHolder,data: T,position: Int)

    private var mOnClickListener: OnItemClickListner? = null
    private var mDatas: MutableList<T>? = null

    init {
        mDatas = ArrayList()
    }

    fun setOnItemClickListener(listner: OnItemClickListner) {
        this.mOnClickListener = listner
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
        val binding = DataBindingUtil.inflate<ViewDataBinding>(LayoutInflater.from(parent.context),layoutId,parent,false)
        val viewHolder = BaseViewHolder(binding.root,binding)
        return viewHolder
    }

    override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
        onBindView(holder,mDatas!!.get(position),position)
        holder?.itemView?.setOnClickListener { view ->
            mOnClickListener?.onItemClick(view,position)
        }
    }

    override fun getItemCount(): Int {
        return mDatas!!.size
    }

    fun addDatas(datas: ArrayList<T>,isWantAnimator: Boolean) {
        val start = mDatas!!.size
        mDatas!!.addAll(datas)
        val end = mDatas!!.size - 1
        if (isWantAnimator)
            notifyItemRangeInserted(start,end)
        else
            notifyDataSetChanged()
    }

    fun addData(data: T,isWantAnimator: Boolean) {
        mDatas!!.add(data)
        if (isWantAnimator)
            notifyItemInserted(mDatas!!.size-1)
        else
            notifyDataSetChanged()
    }

    fun replaceAllData(datas: ArrayList<T>,isWantAnimator: Boolean) {
        mDatas!!.clear()
        mDatas!!.addAll(datas)
        if (isWantAnimator)
            notifyItemRangeInserted(0,mDatas!!.size-1)
        else
            notifyDataSetChanged()
    }

    fun getData(position: Int): T? {
        if (position >= 0 && position < mDatas!!.size) {
            return mDatas!!.get(position)
        }
        return null
    }

}

